src/Effect/Ref.js

```diff
@@ -1,8 +1,10 @@
 "use strict";
 
-exports.new = function (val) {
+exports.self = function (f) {
   return function () {
-    return { value: val };
+    var ref = { value: null };
+    ref.value = f(ref);
```

Comment:
I think it would be safer to throw an exception if the value isn't set yet, otherwise you break type-safety since `null` can silently sneak into your program.

Reply:
How would we do that?

Reply:
You have to guard read/modify, unfortunately.
```js
var EMPTY = {};

exports.self = function (f) {
  return function () {
    var ref = { value: EMPTY };
    ref.value = f(ref);
    return ref;
  };
};

exports.read = function (ref) {
  return function () {
    if (ref.value === EMPTY) {
      throw new Error("Uninitialized ref");
    }
    return ref.value;
  };
};
```

I think this is a reason why I'd rather keep it out of the lib. If a user wants to `unsafeCoerce unit` a value, that's their choice, but I don't want to expose an API that is dangerously and silently unsafe.

Reply:
Can you concoct a scenario in which this thing is unsafe, outside of using more `unsafeCoerce` and such? The `Ref s -> s` function is pure, so it has no opportunity to do anything with the new `Ref` other than use it as a value.

Reply:
The other alternative is to do the `monad-fix` thing of passing in `Unit -> Ref a` instead. By passing in the Ref directly, you have to guard every read, but if you pass in a thunk, you only have to do the guard in the thunk.

Reply:
Oh, I see yes. If the callback is pure, I guess it would be OK. I was thinking it was in Effect.